### PR TITLE
Template build def

### PR DESCRIPTION
**Pull request by goddenrich**
I have written the templating build def it takes an argument substitutions which is a dictionary with the key the build target to the subrule name e.g. '//build:target': 'value' will get the string from file generated by build rule '//build:target_value. This allows us to specify only the target in the file we want to template like we do with the fqn substitution in k8s build defs. NB. this templating only works if the value being substituted in is a single line. I spent some time trying to figure out multiline substitutions but couldn't get it working for now. There are a number of edge cases to consider for special characters so I wrote a little testing script for a yaml file replacement

### CODE REVIEW

File: template/template.build_defs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            deps= value_files,
+        )
+
+    files = filegroup(
+        name=name,
+        srcs=rules,
+        visibility=visibility,
+        labels=labels,
+    )
+    returned_targets['FILES'] = files
```
Comment (peterebden) [inline template/template.build_defs]: this doesn't seem very idiomatic - I would just return the result of the last `filegroup`. Rules don't normally return a dictionary.

Comment (goddenrich) [reply template/template.build_defs]: agreed this was to keep the format the same as the internal k8s build rule which does this but will find a way that doesn't break functionality but is more idiomatic

File: template/template.build_defs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        replacement_targets = [canonicalise(target) for _, target in subs.items()]
+        searches = _search_escaped(subs.keys())
+    elif isinstance(subs, list):
+        replacement_targets = [canonicalise(target) for target in subs]
+        searches = _search_escaped(replacement_targets) # in this case the search string is same as target
+
+    replacements = _replacement_escaped(replacement_targets)
+
+    replace_in_single_quotes = ' '.join([
+        _sed_for_single_quotes(search, replacement) for search, replacement in sorted(zip(searches, replacements))
```
Comment (peterebden) [inline template/template.build_defs]: is there a reason you need `sorted`? the ordering should be deterministic without, but maybe you need them to be explicitly in order?

Comment (goddenrich) [reply template/template.build_defs]: no I think it was legacy from a refactor

**Pull request merged**